suprb/optimizer/rule/ns/ns.py

```diff
@@ -138,171 +120,67 @@ def _optimize(self, X: np.ndarray, y: np.ndarray, n_rules: int) -> list[Rule]:
-        # calculate the 25th percentile value and if it's lower than MNCS_threshold_matched use it instead to filter
-        # a maximum of 25% of the population (to prevent empty populations)
-        maximum_threshold = min(np.percentile([np.count_nonzero(rule.match_) for rule in rules], 
-                                              self.MCNS_threshold_matched), 25)
-        rules = [rule for rule in rules if np.count_nonzero(rule.match_) >= maximum_threshold]
-        return rules
-
-    def _filter_for_progressive_minimal_criteria(self, rules: list[Rule]) -> list[Rule]:
-        threshold_fitness = np.median([rule.fitness_ for rule in rules])
-        rules = [rule for rule in rules if rule.fitness_ >= threshold_fitness]
-        return rules
-
-    def _get_local_score(self, rule: Rule, rules_w_distances: list[tuple[Rule, float]]) -> float:
-        count_worse = 0
-        for x, _ in rules_w_distances[:self.NSLC_threshold]:
-            if x.fitness_ < rule.fitness_:
-                count_worse += 1
-        local_score = count_worse / len(rules_w_distances[:self.NSLC_threshold])
-        return local_score
-
-    def _get_pareto_front(self, rules_with_novelty_score: list[tuple[Rule, float, int]]) -> list[tuple[Rule, float,
-                                                                                                       int]]:
-        rules_with_novelty_score = sorted(rules_with_novelty_score, key=lambda a: (a[1], a[0].fitness_), reverse=True)
-        p_front = [rules_with_novelty_score[0]]
-
-        for r in rules_with_novelty_score[1:]:
-            if r[0].fitness_ >= p_front[-1][0].fitness_:
-                p_front.append(r)
-
-        return p_front
+        return parents
 
-    def _validate_params(self, n_rules: int):
-        if self.ns_type not in ["NS", "NSLC", "MCNS"]:
-            raise ValueError(f"{self.ns_type} is no valid NS-Type.")
-        if self.archive not in ["novelty", "random", "none"]:
-            raise ValueError(f"{self.archive} is no valid Archive-Type.")
-        if self.novelty_fitness_combination not in ["novelty", "50/50", "75/25", "pmcns", "pareto"]:
-            raise ValueError(f"{self.novelty_fitness_combination} is no valid Fitness-Novelty Combination.")
-        if n_rules > (self.mu + self.lmbda):
-            raise ValueError(f"n_rules={n_rules} must be less or equal to mu+lambda={self.mu+self.lmbda}.")
+    def _crossover_and_mutate(self, X: np.ndarray, y: np.ndarray, parents: list[Rule]) -> list[Rule]:
+        children = []
+
+        parent_combinations = zip(*[iter(parents)] * 2)
+        for parent_A, parent_B in parent_combinations:
+            children.extend(self.crossover(A=parent_A, B=parent_B, random_state=self.random_state_))
+
+        children = [self.constraint(self.mutation(child, random_state=self.random_state_)).fit(X, y)
+                    for child in children]
+
+        return children
+
+    def _get_n_best_rules(self, rules: list[Rule], n: int):
```

Comment:
if we use a partially fitness-based objective, shouldn't we incorporate that here?

Reply:
In that case the novelty score gets calculated by the hamming distance and the fitness (in whatever ratio the user has chosen). So using the novelty score here, will already have the fitness incorporated

Reply:
Should we name it something else than novelty score to make this less confusing? Or is it just me?

Reply:
As I understand it, the novelty score is still the correct name here. the fitness is just added into that calculation. So I would keep it as it is